formal-models/README.md

```diff
@@ -112,6 +112,215 @@ configuration:
+its own `ChangeView[1,2,3]` message to exit to the next consensus round (view) if
+something goes wrong in the current one and there's definitely no ability to
+continue consensus process in the current view. Below there's a short description
+of each stage. Please, refer to the model scheme and specification for further
+details.
+
+#### Stage I
+
+Once initialized, consensus node has two ways:
+1. Send its `PrepareRequest`/`PrepareResponse` message (and transmit to the
+  `prepareSent` state).
+2. Decide to go to the next view on timeout or any other valid reason like
+  transaction missing in the node's mempool via sending `ChangeView1` message
+  (and transmit to the `cv1` state).
+
+This scheme is quite similar to the basic dBFT 2.0 model except the new type of
+`ChangeView` message. After that the node enters stage `I` and waits for consensus
+messages of stage `I` (`PrepareRequest` or `PrepareResponse` or `ChangeView1`)
+from at least `M` neighbours which is needed to decide about the next actions.
+The set of received messages can be arranged in the following way:
+
+* `M` messages of `ChangeView1` type denote that `M` nodes have decided to change
+  their view directly after initialization due to invalid/missing `PrepareRequest`
+  which leads to immediate view changing.
+* `M` preparation messages (of type `PrepareRequest` or `PrepareResponse`) with
+  no `ChangeView3` denote that the majority of nodes have decided to commit which
+  denotes the safe transition to stage `II` can be performed and `Commit` message
+  can safely be sent even if there's `ChangeView1` message in the network.
+* `M` messages each of the type `PrepareRequest` or `PrepareResponse` or `ChangeView1`
+  where at least one message is of the type `ChangeView1` denote that at least `M`
+  nodes have reached the stage `I` and the node can safely take further steps.
+  The additional `| Commit | ≤ F ∪ | CV3 | > 0` condition requires the majority of
+  nodes not to have the `Commit` message to be sent so that it's still possible to
+  collect enough `ChangeView[2,3]` messages to change the view in further stages.
+  If so, then the safe transition to stage `II` can be performed and `ChangeView2`
+  message can safely be sent.
+
+#### Stage II
+
+Once the node has `Commit` or `ChangeView2` message sent, it enters the stage `II`
+of the consensus process and waits for at least `M` messages of stage `II`
+(`Commit` or `ChangeView2`) to perform the transition to the next stage. The set
+of accepted messages can be arranged in the following way:
+
+* `M` messages of `ChangeView2` type denote that `M` nodes have decided to change
+  their view directly after entering the stage `II` due to timeout while waiting
+  for the `Commit` messages which leads to immediate view changing.
+* `M` messages of type `Commit` denote that the majority of nodes have decided to
+  commit which denotes the block can be accepted immediately without entering the
+  stage `III`.
```

Comment:
Notice that this is the regular flow of normal dbft 2.0 consensus, it also hasn't changed and proceeds the way it was before.

Reply:
We also need to specify that Commit messages should bear preparation hashes from all nodes (but not including a whole PrepareRequest). This quickly synchronizes nodes still at stage 1 when someone else collects enough preparations. It at the same time prevents malicious/byzantine nodes from sending spoofed commits. The commit message size becomes a little bigger, but since it's just hashes it still fits into a single packet in vast majority of the cases, so it doesn't really matter.